From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Denery <dorofeevij@gmail.com>
Date: Mon, 13 Feb 2023 18:23:18 +0300
Subject: [PATCH] Implement Paper support


diff --git a/build.gradle b/build.gradle
index 01a9ec66c39162b0050118faf744f153f0cb2254..5e0de6eaba8c2f7ea7d9e10e560c2499a95f27ad 100644
--- a/build.gradle
+++ b/build.gradle
@@ -69,6 +69,9 @@ dependencies {
 	api "org.ow2.asm:asm-commons:${project.asm_version}"
 	api "org.ow2.asm:asm-tree:${project.asm_version}"
 	api "org.ow2.asm:asm-util:${project.asm_version}"
+	api("org.sharegov:mjson:1.4.1") {
+		transitive = false
+	}
 
 	api("net.fabricmc:sponge-mixin:${project.mixin_version}") {
 		exclude module: 'launchwrapper'
diff --git a/minecraft/src/main/java/net/fabricmc/loader/impl/game/minecraft/MinecraftGameProvider.java b/minecraft/src/main/java/net/fabricmc/loader/impl/game/minecraft/MinecraftGameProvider.java
index d51321551437bead68505a3c4f45a9af1e0c55e4..21b3996a7ac5282f75c25aebaed630a07e81a50a 100644
--- a/minecraft/src/main/java/net/fabricmc/loader/impl/game/minecraft/MinecraftGameProvider.java
+++ b/minecraft/src/main/java/net/fabricmc/loader/impl/game/minecraft/MinecraftGameProvider.java
@@ -33,6 +33,8 @@ import java.util.Locale;
 import java.util.Map;
 import java.util.Set;
 
+import mjson.Json;
+
 import net.fabricmc.api.EnvType;
 import net.fabricmc.loader.api.ObjectShare;
 import net.fabricmc.loader.api.VersionParsingException;
@@ -438,11 +440,17 @@ public class MinecraftGameProvider implements GameProvider {
 		for (Path lib : miscGameLibraries) {
 			launcher.addToClassPath(lib);
 		}
+		// Toki start - additional Paper libraries
+		Json.read(System.getProperty("toki.libraries")).asJsonList().stream().map(Json::asString).forEach(string -> {
+			final Path libPath = Paths.get(string);
+			launcher.addToClassPath(libPath);
+		});
+		// Toki end
 	}
 
 	@Override
 	public void launch(ClassLoader loader) {
-		String targetClass = entrypoint;
+		String targetClass = System.getProperty("toki.entrypoint"); // Toki - we need org.bukkit.craftbukkit.Main there, but we still need to patch the original server main method in EntrypointPatch.
 
 		if (envType == EnvType.CLIENT && targetClass.contains("Applet")) {
 			targetClass = "net.fabricmc.loader.impl.game.minecraft.applet.AppletMain";
diff --git a/minecraft/src/main/java/net/fabricmc/loader/impl/game/minecraft/patch/EntrypointPatch.java b/minecraft/src/main/java/net/fabricmc/loader/impl/game/minecraft/patch/EntrypointPatch.java
index 8858165dcc7a0cc6fabf0ddd61544f3a07295afc..ace57703416fd91dc1feaf3a66988f35bfab3bc0 100644
--- a/minecraft/src/main/java/net/fabricmc/loader/impl/game/minecraft/patch/EntrypointPatch.java
+++ b/minecraft/src/main/java/net/fabricmc/loader/impl/game/minecraft/patch/EntrypointPatch.java
@@ -107,7 +107,7 @@ public class EntrypointPatch extends GamePatch {
 
 		if (gameEntrypoint == null) {
 			// main method searches
-			MethodNode mainMethod = findMethod(mainClass, (method) -> method.name.equals("main") && method.desc.equals("([Ljava/lang/String;)V") && isPublicStatic(method.access));
+			MethodNode mainMethod = findMethod(mainClass, (method) -> method.name.equals("main") && method.desc.equals("(Ljoptsimple/OptionSet;)V") && isPublicStatic(method.access)); // Toki - change main method
 
 			if (mainMethod == null) {
 				throw new RuntimeException("Could not find main method in " + entrypoint + "!");
@@ -254,7 +254,7 @@ public class EntrypointPatch extends GamePatch {
 				}
 			}
 		} else {
-			gameMethod = findMethod(mainClass, (method) -> method.name.equals("main") && method.desc.equals("([Ljava/lang/String;)V") && isPublicStatic(method.access));
+			gameMethod = findMethod(mainClass, (method) -> method.name.equals("main") && method.desc.equals("(Ljoptsimple/OptionSet;)V") && isPublicStatic(method.access)); // Toki - change main method
 		}
 
 		if (gameMethod == null) {
@@ -306,7 +306,7 @@ public class EntrypointPatch extends GamePatch {
 				// If we do not find this, then we are certain this is 20w22a.
 				MethodNode serverStartMethod = findMethod(mainClass, method -> {
 					if ((method.access & Opcodes.ACC_SYNTHETIC) == 0 // reject non-synthetic
-							|| method.name.equals("main") && method.desc.equals("([Ljava/lang/String;)V")) { // reject main method (theoretically superfluous now)
+							|| method.name.equals("main") && method.desc.equals("(Ljoptsimple/OptionSet;)V")) { // reject main method (theoretically superfluous now) // Toki - change main method to fit cb patch
 						return false;
 					}
 
@@ -396,7 +396,7 @@ public class EntrypointPatch extends GamePatch {
 								return false;
 							}
 
-							return constructorType.getArgumentTypes()[0].getDescriptor().equals("Ljava/lang/Thread;");
+							return constructorType.getArgumentTypes()[0].getDescriptor().equals("Ljoptsimple/OptionSet;") && constructorType.getArgumentTypes()[2].getDescriptor().equals("Ljava/lang/Thread;"); // Toki - change constructor finding
 						}
 
 						return false;
diff --git a/settings.gradle b/settings.gradle
deleted file mode 100644
index 9c7c525d2c495a109c6fccf11bf6a4123d0e38c8..0000000000000000000000000000000000000000
--- a/settings.gradle
+++ /dev/null
@@ -1,18 +0,0 @@
-pluginManagement {
-	repositories {
-		maven {
-			url = "https://maven.fabricmc.net"
-			name = "FabricMC"
-		}
-		gradlePluginPortal()
-	}
-}
-rootProject.name='fabric-loader'
-
-include "minecraft"
-
-if (JavaVersion.current().isCompatibleWith(JavaVersion.VERSION_17)) {
-	include "minecraft:minecraft-test"
-} else {
-	println("Minecraft test sub project requires java 17 or higher!")
-}
\ No newline at end of file
diff --git a/src/main/resources/fabric-installer.json b/src/main/resources/fabric-installer.json
index 01d3601816b0d65ab4eec3f936de22b2868f0a0a..e0ff157c9af93575c81fa8ddee3379771593d1d2 100644
--- a/src/main/resources/fabric-installer.json
+++ b/src/main/resources/fabric-installer.json
@@ -39,6 +39,10 @@
       {
         "name": "org.ow2.asm:asm-util:9.4",
         "url": "https://maven.fabricmc.net/"
+      },
+      {
+        "name": "org.sharegov:mjson:1.4.1",
+        "url": "https://repo.maven.apache.org/maven2/"
       }
     ],
     "server": [
